src/service.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent pebble service module."""
+
+import logging
+import os
+import pwd
+import time
+from pathlib import Path
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+from jinja2 import Template
+
+from charm_state import State
+
+logger = logging.getLogger(__name__)
+AGENT_SERVICE_NAME = "jenkins-agent"
+AGENT_PACKAGE_NAME = "jenkins-agent"
+SYSTEMD_SERVICE_CONF_DIR = "/etc/systemd/system/jenkins-agent.service.d/"
+PPA_GPG_KEY_ID = "67393A94A577DC24"
+READINESS_CHECK_DELAY = 30
+
+
+class PackageInstallError(Exception):
+    """Exception raised when package installation fails."""
+
+
+class ServiceRestartError(Exception):
+    """Exception raised when failing to start the agent service."""
+
+
+class JenkinsAgentService:
+    """Jenkins agent service class.
+
+    Attrs:
+       is_active: Indicate if the agent service is active and running.
+    """
+
+    def __init__(self, state: State):
+        """Initialize the jenkins agent service.
+
+        Args:
+            state: The Jenkins agent state.
+        """
+        self.state = state
+
+    def _render_file(self, path: str, content: str, mode: int) -> None:
+        """Write a content rendered from a template to a file.
+
+        Args:
+            path: the path to the file.
+            content: the data to be written to the file.
+            mode: access permission mask applied to the
+              file using chmod (e.g. 0o640).
+        """
+        with open(path, "w+", encoding="utf-8") as file:
+            file.write(content)
+        # Ensure correct permissions are set on the file.
+        os.chmod(path, mode)
+        try:
+            # Get the uid/gid for the root user (running the service).
+            # TODO: the user running the jenkins agent is currently root
+            # we should replace this by defining a dedicated user in the apt package
+            u = pwd.getpwnam("root")
+            # Set the correct ownership for the file.
+            os.chown(path, uid=u.pw_uid, gid=u.pw_gid)
+        except KeyError:
+            # Ignore non existing user error when it wasn't created yet.
+            pass
+
+    @property
+    def is_active(self) -> bool:
+        """Indicate if the jenkins agent service is active."""
+        return systemd.service_running(AGENT_SERVICE_NAME)
```

Comment:
Is there any possibility of this raising an error?

Reply:
Good point, It can actually raises `SystemError` if the subprocess call to `systemctl` fail for whatever reason. I added error handling to output a log message and return `false` when an exception is caught.